ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topkdistinct/TopkDistinctKudaf.java

```diff
@@ -20,119 +20,81 @@
   TopkDistinctKudaf(final String functionName,
                     final int argIndexInValue,
                     final int tkVal,
                     final Schema outputSchema,
                     final Class<T> ttClass) {
     super(
-        functionName,
-        argIndexInValue,
-            () -> (T[]) Array.newInstance(ttClass, tkVal),
+        functionName, argIndexInValue,
+        (() -> new ArrayList<T>()),
         SchemaBuilder.array(outputSchema).build(),
         Collections.singletonList(outputSchema)
     );
 
     this.tkVal = tkVal;
     this.ttClass = ttClass;
     this.outputSchema = outputSchema;
-    this.comparator = (v1, v2) -> {
-      if (v1 == null && v2 == null) {
-        return 0;
-      }
-      if (v1 == null) {
-        return 1;
-      }
-      if (v2 == null) {
-        return -1;
-      }
-
-      return Comparator.<T>reverseOrder().compare(v1, v2);
-    };
   }
 
   @Override
-  public T[] aggregate(final T currentValue, final T[] aggregateValue) {
-    if (currentValue == null) {
-      return aggregateValue;
-    }
-
-    final T last = aggregateValue[aggregateValue.length - 1];
-    if (last != null && currentValue.compareTo(last) <= 0) {
-      return aggregateValue;
-    }
-
-    if (ArrayUtil.containsValue(currentValue, aggregateValue)) {
-      return aggregateValue;
+  public ArrayList<T> aggregate(final T currentVal, final ArrayList<T> currentAggValList) {
+    if (currentVal == null) {
+      return currentAggValList;
     }
 
-    final int nullIndex = ArrayUtil.getNullIndex(aggregateValue);
-    if (nullIndex != -1) {
-      aggregateValue[nullIndex] = currentValue;
-      Arrays.sort(aggregateValue, comparator);
-      return aggregateValue;
+    Set<T> set = new HashSet<>(currentAggValList);
+    set.add(currentVal);
+    ArrayList list = new ArrayList(set);
+    Collections.sort(list);
+    Collections.reverse(list);
+    if (list.size() > tkVal) {
+      list.remove(list.size() - 1);
     }
-
-    aggregateValue[aggregateValue.length - 1] = currentValue;
-    Arrays.sort(aggregateValue, comparator);
-    return aggregateValue;
+    return list;
   }
 
   @SuppressWarnings("unchecked")
   @Override
-  public Merger<String, T[]> getMerger() {
-    return (aggKey, aggOne, aggTwo) -> {
-      final T[] merged = (T[]) Array.newInstance(ttClass, tkVal);
-
-      int idx1 = 0;
-      int idx2 = 0;
-      for (int i = 0; i != tkVal; ++i) {
-        final T v1 = idx1 < aggOne.length ? aggOne[idx1] : null;
-        final T v2 = idx2 < aggTwo.length ? aggTwo[idx2] : null;
-
-        final int result = comparator.compare(v1, v2);
-        if (result < 0) {
-          merged[i] = v1;
-          idx1++;
-        } else if (result == 0) {
-          merged[i] = v1;
-          idx1++;
-          idx2++;
-        } else {
-          merged[i] = v2;
-          idx2++;
-        }
+  public Merger<String, ArrayList<T>> getMerger() {
+    return (aggKey, aggOneList, aggTwoList) -> {
```

Comment:
Again, performance concerns here.

Reply:
I keep this as is for now until rewrite of both TopK and TopKDistinct with struct.